File: src/bmsx_cpp/machine/runtime/save_state.cpp
#include "machine/runtime/save_state.h"

#include "machine/runtime/cpu_state.h"
#include "machine/runtime/save_machine_state.h"
#include "machine/runtime/runtime.h"
#include "render/vdp/context_state.h"

namespace bmsx {

RuntimeSaveState captureRuntimeSaveState(Runtime& runtime) {
	RuntimeSaveState state;
	state.machineState = captureRuntimeSaveMachineState(runtime);
	state.cpuState = captureRuntimeCpuState(runtime);
	state.systemProgramActive = !runtime.m_cartProgramStarted;
	state.luaInitialized = runtime.m_luaInitialized;
	state.luaRuntimeFailed = runtime.m_runtimeFailed;
	state.randomSeed = runtime.m_randomSeedValue;
	state.pendingEntryCall = runtime.m_pendingCall == Runtime::PendingCall::Entry;
	return state;
}

void applyRuntimeSaveState(Runtime& runtime, const RuntimeSaveState& state) {
	if (state.systemProgramActive) {
		runtime.enterSystemFirmware();
	} else {
		runtime.enterCartProgram();
	}
	applyRuntimeSaveMachineState(runtime, state.machineState);
	restoreVdpContextState(runtime.machine.vdp, runtime.view());
	applyRuntimeCpuState(runtime, state.cpuState);
	runtime.m_luaInitialized = state.luaInitialized;
	runtime.m_runtimeFailed = state.luaRuntimeFailed;
	runtime.m_randomSeedValue = state.randomSeed;
	runtime.m_pendingCall = state.pendingEntryCall ? Runtime::PendingCall::Entry : Runtime::PendingCall::None;
}

} // namespace bmsx

File: src/bmsx/machine/runtime/save_state.ts
import type { CpuRuntimeState } from '../cpu/cpu';
import { consoleCore } from '../../core/console';
import type { RuntimeSaveMachineState } from './save_machine_state';
import type { Runtime } from './runtime';
import { applyRuntimeCpuState, captureRuntimeCpuState } from './cpu_state';
import { restoreVdpContextState } from '../../render/vdp/context_state';
import { applyRuntimeSaveMachineState, captureRuntimeSaveMachineState } from './save_machine_state';

export type RuntimeSaveState = {
	machineState: RuntimeSaveMachineState;
	cpuState: CpuRuntimeState;
	systemProgramActive: boolean;
	luaInitialized: boolean;
	luaRuntimeFailed: boolean;
	randomSeed: number;
	pendingEntryCall: boolean;
};

export function captureRuntimeSaveState(runtime: Runtime): RuntimeSaveState {
	return {
		machineState: captureRuntimeSaveMachineState(runtime),
		cpuState: captureRuntimeCpuState(runtime),
		systemProgramActive: !runtime.cartProgramStarted,
		luaInitialized: runtime.luaInitialized,
		luaRuntimeFailed: runtime.luaRuntimeFailed,
		randomSeed: runtime.randomSeedValue,
		pendingEntryCall: runtime.pendingCall === 'entry',
	};
}

export function applyRuntimeSaveState(runtime: Runtime, state: RuntimeSaveState): void {
	if (state.systemProgramActive) {
		runtime.enterSystemFirmware();
	} else {
		runtime.enterCartProgram();
	}
	applyRuntimeSaveMachineState(runtime, state.machineState);
	restoreVdpContextState(runtime.machine.vdp, consoleCore.view);
	applyRuntimeCpuState(runtime, state.cpuState);
	runtime.luaInitialized = state.luaInitialized;
	runtime.luaRuntimeFailed = state.luaRuntimeFailed;
	runtime.randomSeedValue = state.randomSeed;
	runtime.pendingCall = state.pendingEntryCall ? 'entry' : null;
}
